### ISSUE TITLE
Persisted queries and better access to cache

### ISSUE THREAD

**Issue by adjourn** [grafoojs/grafoo#74]
Im really intrigued by this library, mainly because of small size (Apollo is close to 100Kb before gzip and that's without `graphql`) and simple core API to build my own thing with (e.g. on React hooks). Great job!

However I have few **feature requests**,  I'd like to hear your opinion before even thinking about switching:

**1.** What do you think about reworking `flush` or adding a new API to access cache in app? Word `flush` seems odd in that context. Few options: method to read data by key or even expose map directly.

Direct access opens up so many possibilities which enables us to build pretty complex behaviors with these primitives, e.g. show partial data while loading when navigating from list view to detail view. I just tested it and 2 identical queries except with few different fields don't return the same item from cache (not sure if `partial` bug or intended, I never understood Apollo `partial` either).

**2.** Option to generate `id` for query objects and pass it to fetch function, then we're able to only send query `id` instead of whole query, less bytes to upload and ability to white-list (persisted queries). Preferably a hash based on query string, then we don't end up with same queries and different `ids`.

Cheers

**Comment by malbernaz** [comment]
Hey @adjourn, thanks for your feedback!

In client you have the `read` method that takes as parameters the query and variables. That's how you read from the cache, `flush` is not meant to be used for that matter. A suggestion if you are using the react or preact integration is to use the client directly, which is exposed in the `Consumer` render function.

Now, if you have an idea of how this API could be like, I'm all ears. It would be rather easier, really, to grasp what kind of modifications I should have to do on `core` internals and answer you if it's doable or not.

I'm assuming that your primary problem is that you want to access queries partially. You are right about the fact that queries with different fields are not the same in the cache. That was a thoughtful decision when I designed how the library would work, because the logic I came at the time would if added would exceed my bundle size budget.

For now my suggestion is that you try to gather your queries in your top level components and pass then down via props or context.

But I'll revisit this topic and give you an answer soon.

Persisted queries: I'm all in! And i think it's very doable. I'll do my research here to see what it takes.

**Comment by adjourn** [comment]
Im planning to build my own React integration if I decide to switch (which I probably will, I really like this) because I have very specific demands for that API. But for core API I was thinking something like this:
     
```js
// example usage
const item = client.readByKey('123'); // idFields: ['id']
const item = client.readByKey('123User'); // idFields: ['id', '__typename']
```

```js
// example implementation
function readByKey(key) {
  // returns item if exists,
  // otherwise undefined
  return objectsMap[key];
}
```
    
It's pretty clear that `flush` is not meant to be used like that, name pretty much says it all but I mentioned it because it's currently the only way to access items by key. And to be clear, Im not super interested in the feature to merge same objects (if it adds a lot to bundle), if there was a way to get items by key, pretty much all my use cases would be covered. This is where other users' input would be great.

One of my example use case:
1. Fetch and render list of items (`{items: { id, title }}`)
2. User clicks on one item, route changes to `/item/123`
3. Build the `key` from url (e.g `item` is type, `123` is id, key is `123Item`)
3. Use `client.readByKey(key)` to read partial item from cache
4. If exists: show partial data and small spinner, doesn't exists: show big spinner
5. Fetch single item to get rest of the data (`{item(id: "123") { id, title, desc, comments }}`),
could even use 2nd query (`{item(id: "123") { desc, comments }}`) if partial data exists

Other way would be to just `return { execute, listen, write, read, flush, reset, objectsMap };` and let us use it to our heart's content but it kind of exposes internals and we probably don't want that.

**Comment by malbernaz** [comment]
It looks like a good idea! If you'd like to give it a shot with a PR I can support you, on the contrary I can land this maybe next week. I'll open an issue, just to let it mapped.

Out of topic I'm interested to know about your needs regarding the react integration. What are the demands you have that are unmet by `@grafoo/react`? Any feedback on that would be really useful.

**Comment by malbernaz** [comment]
Any thought on that @mogelbrod?

**Comment by adjourn** [comment]
Lerna and TypeScript are new technologies for me. Would love to pick them up at some point in order to contribute something meatier, maybe you could add this little feature in order to get it merged a lot sooner?

I also suggest to add small contribution guide to readme (what commands to run in what order, etc).

---  

> Out of topic I'm interested to know about your needs regarding the react integration. What are the demands you have that are unmet by `@grafoo/react`? Any feedback on that would be really useful.

I've never liked render props because it creates a false hierarchy in code (indentation) and I try to avoid nesting where I can. It also adds a lot of noise to dev tools. Matter of preference I guess. Now that React has hooks, I use them where I can, including data fetching.

Other than that, I really don't have any other demands that generic library could meet: I plan to write data hooks coupled with internal systems like global state and custom behaviors like query hook that returns partial data (using previously mentioned `readByKey`) if `loading === true`.

**Comment by mogelbrod** [comment]
Sounds good, especially the partial cache fetching! A `resetByKey()` would probably be nice to have along with the `readByKey()`.

**Comment by mogelbrod** [comment]
Agreed on the contribution guide as well - it took an hour or so to get a working dev environment up for this repository. From what I can remember I did the following (but they're likely suboptimal):

```sh
npm i # install deps
lerna bootstrap # initial setup
```

Whenever I changed something the only workflow that I could get running was:
```sh
lerna bootstrap # rebuild
lerna run test # run test (or another script)
```

I wasn't able to run tests from within a single package as `babel` and `grafoo-bundle` wasn't in the runtime path.

**Comment by mogelbrod** [comment]
Adding `resetByKey()` (which clears the cache for a key otherwise accessible by `readByKey()`) would keep the reset API in parity with the `read*()` variants (with the addition of `readByKey()`).

**Comment by malbernaz** [comment]
It removes an item with a given key from the cache?

I agree that clear is a better term then rest.

**Comment by adjourn** [comment]
More cranual reset would be very handy, these are exactly the primitives I want from libraries to build app specific behaviors and not bloat library with edge case features.

@mogelbrod I agree, `reset*` doesn't sound good on item level. `delete*` would be another option.

But should it be`*ByKey(key)` or `*ByQuery(query, vars)`? Both? Or one method and let it check if arg is `String`? I can already think of multiple use cases I'd like to implement like cache invalidation for example:

```js
// pesudo code (React hook)

const expirationMap = { 
  // query.id could be used here if 
  // it gets added (persisted queries)
  // [query.id]: expirationEpoch, 
  fdgf43oij54: 1549626777
};

function useExpiringData(query, vars, duration) {
  const client = useContext(GrafooContext);

  // e.g if no duration or 0, bust cache
  if (!duration) {
    // - delete item from cache by query if it exists
    // - execute, goes to network
    // - write to cache
    // - don't update expirationMap
  }

  if (expirationMap[query.id]) {
    const now = (new Date).getTime();
    if (expirationMap[query.id] < now) {
      // CACHED, EXPIRED
      // - delete item from cache by query
      // - execute again, goes to network
      // - write to cache
      // - update expirationMap (using duration)
    } else {
      // CACHED, NOT EXPIRED
      // execute (returns from cache)
    }
  } else {
    // NOT CACHED
    // - execute, goes to network
    // - write to cache
    // - update expirationMap (using duration)
  }
}
```

**Comment by mogelbrod** [comment]
clear/delete/remove are all possible options for the cache removal, but `reset` no longer feels correct when extended into `*ByKey()` like you wrote. `*ByQuery()` is a potential future extension which I think should be its own function since they have pretty different behaviors.

👍 for cache expiration, I was thinking of that myself.
